design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+- SLA - Service Level Agreement - what are the SLOs written into an agreement between services (or Red Hat and end users for that matter).
+- Alert - A warning triggered when metrics went beyond a level defined in SLO.
+[Here’s the OpenShift Dedicated SLA (go to page 9)](https://www.redhat.com/licenses/Appendix_4_Online_Global_20200406.pdf).
+- Prometheus - A time-series database. We store metrics there.
+- Grafana - A service that presents diagrams based on metrics from Prometheus.
+- EFK Stack - Elsticsearch, Fluentd and Kibana. A typical log-collecting stack.
+- CLO - Cluster Logging Operator
+- CR - Custom Resource. An API provided by Keycloak Operator for the context of this document.
+
+## Motivation
+
+Managing Keycloak cluster requires involving a dedicated SRE (Site Reliability Engineering) Team. Such a team needs to
+be able to:
+
+* Diagnose if Keycloak cluster is up and running
+* Have sufficient information to diagnose issues
+* Have enough data and metrics to foresee problems that will happen in the near future
+
+Typically, an SRE Team uses dashboards and alerts. Both of them are created based on metrics exported by a Service
+(in this case, the Service is both Keycloak and Keycloak Operator).
+
+Based on metrics, dashboards and alerts, it is possible to define SLA - an agreement between a Service Provider and a
+Service Consumer.
+
+## Scope
+
+What's in the scope of this design:
+- Designing a new Healcheck API. This has been covered in a separate design document. See [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+- Designing and implementing a dashboard for the SRE Team.
+- Defining SLIs and SLOs as well as proposing template SLAs for a managed Keycloak cluster.
+- Defining metrics required by the proposed dashboard as well as designing a mechanism for exporting them to prometheus.
+- Design logging improvements for better error tracking.
+
+What's out of the scope:
+- Traceability - tracking a user across the whole authorization flow. At the moment we don't have any requirements
+in this direction. If we get such a requirement, we will track this progress in a separate epic and design.
+
+## SRE Dashboard
+
+![SRE Dashboard](img/observability-dashboard.png)
+
+NOTE: The diagram implementation and specific charts might change during the implementation phase. The above proposal
+shows only the general idea.
+
+The proposal consists of 3 sections:
+- System health - Enables the SRE Team to quickly tell if Keycloak cluster is up and running. This section consists of:
+  - Readiness and Liveness Probes chart
+  - HTTP 5xx error rate for all major Keycloak Services, including: Admin Console, Account Console, OIDC, SAML
+  - Aggregated latency across the most important endpoints
+  - (Optional) Error rate in logs [^1]
+- Detailed Metrics - This section helps foresee upcoming failures. It includes:
+  - Pod memory, CPU and container storage charts
+  - GC Times
+  - Heap and non-heap memory metrics
```

Comment:
Detailed Metrics - What about system resources, or something like that instead?

Can we think about a way to make metrics understandable to a non-Java guy?

Reply:
For us it would also be interesting to combine this view with the kubernetes statistics. So one would have the internal and the external view on the workload. For example if applied resource limits for cpu and memory could be made visible within the same view

Reply:
@stianst The purpose of the "Detailed Metrics" is to provide the SRE Team more detailed information. This is very useful when reporting JIRAs against our team. We need to go a bit further than typical OS metrics here to make such reports more valuable. Keycloak is a Java-based app. The SRE Team needs to have some basic training in monitoring JVM-based apps. I don't see other way.

@bs-atil Putting Resources and Limits in the same graph seems like a good idea. Note, that Pod Memory and CPU metrics are reported by OpenShift/Kubernetes directly from CGroups. Those are the once we are interested in on left side of the dashboard (Pod Memory, CPU and Storage). The right side is more JVM-related. There you have Heap/non-heap and GC times. Perhaps it makes sense to put Kubernetes Resources in Heap/non-heap graph as well. I'll try to incorporate this into the sketch diagram and see how it looks like.

Reply:
Added both Resources and Limits to the graphs.